File: picassoboxes.php
<?php
class picassobox {
	public function __construct($box, $colour) {
		$this->dim = count($box);
		$this->box = $box;
		$this->colour = $colour;
		$this->sides = [];
		$this->mass = 1;
		foreach($box as $side) {
			$dx = $side[1]-$side[0];
			$this->sides[] = $dx;
			$this->mass = $this->mass*$dx;
		}
		return;
	}

	public function randomunif() {
		$x = [];
		foreach($this->box as $xx) {
			$u = mt_rand(0,2147483647-1)/2147483647;
			$x[] = $xx[0] + $u*($xx[1]-$xx[0]);
		}
		return $x;
	}

	public function point($coord) {
		if(!is_array($coord)) $coord = array_fill(0, $this->dim, $coord);
		$x = [];
		foreach($this->box as $k => $xx) {
			$u = $coord[$k];
			$x[] = $xx[0] + $u*($xx[1]-$xx[0]);
		}
		return $x;
	}

	public function intersects($PB) {
		$bool = TRUE;
		foreach($this->box as $k => $side) {
			$side_ = $PB->box[$k];
			if($side_[1] <= $side[0] or $side[1] <= $side_[0]) {
				$bool = FALSE;
				break;
			}
		}	
		return $bool;
	}

	public function subseteq($PB) {
		$bool = TRUE;
		foreach($this->box as $k => $side) {
			$side_ = $PB->box[$k];
			if($side_[0] <= $side[0] and $side[0] <= $side_[1] and $side[1] <= $side_[1]) continue;
			$bool = FALSE;
			break;
		}
		return $bool;
	}

	public function modify($params) {
		$X = new picassobox($this->sides, $this->colour);

		if(isset($params['preshift'])) for($k=0; $k<$this->dim; $k++) $X->sides[k][0] += $params['preshift'][k];
		if(isset($params['postshift'])) for($k=0; $k<$this->dim; $k++) $X->sides[k][1] += $params['postshift'][k];
		if(isset($params['colour'])) $X->colour = $params['colour'];

		return $X;
	}

	public function mince($PBfilt, $col=1) {
		if(!$this->intersects($PBfilt)) return [new picassobox($this->box, 0)];

		$sides = [];
		$trace = [];
		foreach($this->box as $k => $side) {
			$trace_ = [];
			$sides_ = [];
			$side_ = $PBfilt->box[$k];
			$a = $side[0];
			$b = $side[1];
			$aa = $side_[0];
			$bb = $side_[1];

			if($aa < $a) $aa = $a;
			if($bb > $b) $bb = $b;
			if($bb < $aa) $bb = $aa;

			if($a < $aa) {
				$sides_[] = [$a,$aa];
				$trace_[] = 0;
			}
			if($aa < $bb) {
				$sides_[] = [$aa,$bb];
				$trace_[] = 1;
			}
			if($bb < $b) {
				$sides_[] = [$bb,$b];
				$trace_[] = 0;
			}

			$sides[] = $sides_;
			$trace[] = $trace_;
		}

		$traces = picassobox::cartproduct($trace);
		$part = picassobox::cartproduct($sides);
		$boxes = [];
		foreach($traces as $i => $trace) {
			$box = $part[$i];
			$__col = 0;
			if(array_product($trace) === 1) $__col = $col;
			$PB = new picassobox($box,$__col);
			$boxes[] = $PB;
		}
		return $boxes;
	}

	private static function cartproduct($liste) {
		$n = count($liste);
		if($n === 0) return [];
		$prod = [];
		if($n === 1) {
			foreach($liste[0] as $x) $prod[] = [$x];
		} else {
			$prod_ = picassobox::cartproduct(array_slice($liste, 1));
			foreach($prod_ as $y) foreach($liste[0] as $x) $prod[] = array_merge([$x], $y);
		}
		return $prod;
	}
}


class picassoboxes {
	private $filter = [];

	public function __construct($box) {
		$box = picassoboxes::clonelist($box);
		$grenze_lo = [];
		$grenze_hi = [];
		foreach($box as $i => $x) {
			if(!is_array($x)) $box[$i] = [0, $x];
			$grenze_lo[] = $x[0];
			$grenze_hi[] = $x[1];
		}

		$this->dim = count($box);
		$this->box = new picassobox($box, 0);
		$this->mass = $this->box->mass;
		$this->boxes = [];
		$this->bounds = [];

		foreach(range(0, $this->dim - 1) as $k) {
			$box_ = picassoboxes::clonelist($box);
			$box_[$k][0] = $box[$k][1];
			$this->bounds[] = new picassobox($box_, 1);
		}

		$this->buffer = array_fill(0, $this->dim, 0);
		return;
	}

	public function getpartition() {
		$part = [$this->box];
		$boxes_ = [];
		foreach($this->boxes as $PB) $boxes_[] = $PB;
		foreach($this->bounds as $PB) $boxes_[] = $PB;
		foreach($boxes_ as $PB) {
			$shift = [];
			foreach($this->bufferdim as $k => $val) $shift[] = -$val;
			$PB = $PB->modify(['preshift'=>$shift, 'colour'=>1]);
			$part_ = pyArray();
			foreach($part as $k => $PB_) {
				if($PB_->colour == 0) {
					$part__ = $PB_->mince($PB);
				} else {
					$part__ = [$PB_];
				}
				foreach($part__ as $PB__) $part_[] = $PB__;
			}
			$part = $part_;
		}
		return($part);
	}

	public function randomselection($depth=0, $PB=NULL) {
		if($depth == 0) return($this->randomselection(1, $this->box));
		if($depth > count($this->filter)) return($PB);

		$PB_e = NULL;
		$part = $PB->mince($this->filter[$depth]);
		$part = array_filter($part, function($PB_) {
			return($PB_->colour == 0 && $PB_->mass > 0);
		});
		if(count($part) == 0) return(NULL);

		$wt = [];
		foreach($part as $PB_) $wt[] = $PB_->mass;
		$wt_ = $wt;
		$len_ind = count($part);
		$ind = range(0, $len_ind-1);
		while($len_ind > 0) {
			$F = [0];
			$sum = 0;
			foreach($wt_ as $m) {
				$sum += $m;
				$F[] = $sum;
			}
			$u = mt_rand(0,2147483647-1)/2147483647;
			$m = $sum*$u;
			$j = 0;
			for($i=0; $i<$len_ind; $i++) {
				if($m >= F[$i+1]) continue;
				$j = $i;
				break;
			}
			$PB_ = $part[$ind[$j]];
			$PB_e = $this->randomselection($depth+1, $PB_);
			if($PB_e instanceof picassobox) break;
			$ind_ = [];
			for($i=0; $i<$len_ind; $i++) {
				if($i === $j) continue;
				$ind_[] = $ind[$i];
				$wt_[] = $wt_[$i];
			}
			$len_ind--;
		}
		return($PB_e);
	}

	public function addrandomcell($bufferdim, $n=1) {
		if(is_numeric($bufferdim)) $bufferdim = array_fill($bufferdim, $this->dim, $bufferdim);
		if(!is_array($bufferdim)) $bufferdim = array_fill(0, $this->dim, 0);

		$this->buffer = $bufferdim;

		if(n == 0) {
			return([]);
		} else if($n > 1) {
			$cells = [];
			for($i=0; $i<$n; $i++) {
				$cell = $this->addrandomcell($bufferdim, 1);
				if(!($cell instanceof picassobox)) break;
				$cells[] = $cell;
			}
			return(cells);
		}

		# Partition für zulässige Teile neu berechnen:
		$this->filter = [];
		$shift = [];
		foreach($bufferdim as $k => $val) $shift[] = -$val;
		foreach($this->bounds as $PB) $this->filter[] = $PB->modify(['preshift'=>$shift, 'colour'=>1]);
		foreach($this->boxes as $PB) $this->filter[] = $PB->modify(['preshift'=>$shift, 'colour'=>1]);

		# Zufällige Selektion eines zulässigen Teils der Partition:
		$PB_part = $this->randomselection();
		if(!($PB_part instanceof picassobox)) return(NULL);

		# Zufällige Selektion eines Punkts in Box:
		$x = $PB_part->randomunif();
		$sides = [];
		foreach($bufferdim as $k => $h) $sides[] = [$x[$k], $x[$k]+$h];
		$PB = new picassobox($sides, 1);

		# Füge zum Netzwerk von Zellen:
		$this->boxes[] = $PB;

		return($PB);
	}

	private static function clonelist($x, $deep=TRUE) {
		$xx = $x;
		if(is_array($x)) {
			$xx = [];
			foreach($x as $y) {
				$yy = $y;
				if($deep) $yy = picassoboxes::clonelist($y, TRUE);
				$xx[] = $yy;
			}
		} else if(is_object($x)) {
			$xx = (object) [];
			foreach($x as $key) {
				$y = $x->{$key};
				$yy = $y;
				if($deep) $yy = picassoboxes::clonelist($y, TRUE);
				$xx->{$key} = $yy;
			}
		}
		return $xx;
	}
}
?>

File: picassoboxes.py
import numpy;


class picassobox:
	def __init__(self, sides, colour):
		self.dim = len(sides);
		self.sides = sides;
		self.colour = colour;
		self.mass = 1;
		for side in sides:
			dx = side[1]-side[0];
			self.mass = self.mass*dx;
		return;

	def randomunif(self):
		x = [];
		for xx in self.sides:
			u = numpy.random.uniform(0,1);
			x.append(xx[0] + u*(xx[1]-xx[0]));
		return x;

	def point(self, coord):
		if not isinstance(coord, list):
			coord = [coord for k in range(0, self.dim)];
		x = [];
		for k, xx in enumerate(self.sides):
			u = coord[k];
			x.append(xx[0] + u*(xx[1]-xx[0]));
		return x;

	def intersects(self, PB):
		bool = True;
		for k,side in enumerate(self.sides):
			side_ = PB.sides[k];
			if side_[1] <= side[0] or side[1] <= side_[0]:
				bool = False;
				break;
		return bool;

	def subseteq(self, PB):
		bool = True;
		for k, side in enumerate(self.sides):
			side_ = PB.sides[k];
			if side_[0] <= side[0] and side[0] <= side_[1] and side[1] <= side_[1]:
				continue;
			bool = False;
			break;
		return bool;

	def modify(self, params):
		X = picassobox(self.sides, self.colour);

		if 'preshift' in params:
			for k  in range(0, self.dim):
				X.sides[k][0] += params['preshift'][k];
		if 'postshift' in params:
			for k  in range(0, self.dim):
				X.sides[k][1] += params['postshift'][k];
		if 'colour' in params:
			X.colour = params['colour'];

		return X;

	def mince(self, PBfilt, col=1):
		if not self.intersects(PBfilt):
			return [picassobox(self.sides, 0)];

		sides = [];
		trace = [];
		for k,side in enumerate(self.sides):
			trace_ = [];
			sides_ = [];
			side_ = PBfilt.sides[k];
			a = side[0];
			b = side[1];
			aa = side_[0];
			bb = side_[1];

			if aa < a:
				aa = a;
			if bb > b:
				bb = b;
			if bb < aa:
				bb = aa;

			if a < aa:			
				sides_.append([a,aa]);
				trace_.append(0);
			if aa < bb:
				sides_.append([aa,bb]);
				trace_.append(1);
			if bb < b:
				sides_.append([bb,b]);
				trace_.append(0);

			sides.append(sides_);
			trace.append(trace_);
			pass;

		part = self.cartproduct(sides);
		traces = self.cartproduct(trace);
		boxes = [];
		for i,trace in enumerate(traces):
			sides = part[i];
			col_ = 0;
			if numpy.prod(trace) == 1:
				col_ = col;
			PB = picassobox(sides, col_);
			boxes.append(PB);
		return boxes;

	def cartproduct(self, liste):
		n = len(liste);
		if n == 0:
			return [];
		prod = [];
		if n == 1:
			for x in liste[0]:
				prod.append([x]);
		else:
			prod_ = self.cartproduct(liste[1:]);
			for y in prod_:
				for x in liste[0]:
					prod.append([x]+y);
		return prod;

	pass;


class picassoboxes:
	def __init__(self, sides):
		sides = self.clonelist(sides);
		for i, x in enumerate(sides):
			if not isinstance(x, list):
				sides[i] = [0, x];

		self.dim = len(sides);
		self.box = picassobox(sides, 0);
		self.boxes = [];
		self.bounds = [];

		for k in range(0, self.dim):
			sides_ = self.clonelist(sides);
			sides_[k][0] = sides[k][1];
			self.bounds.append(picassobox(sides_, 1));

		self.buffer = [0 for i in range(0, self.dim)];
		return;

	def getboxes(self):
		return [PB.sides for PB in self.boxes];

	def getpartition(self):
		part <- [self.box];
		for PB in self.boxes + self.bounds:
			PB = PB.modify(preshift=-self.buffer, colour=1);
			part_ = [];
			for PB_ in part:
				if PB_.colour == 0:
					part__ = PB_.mince(PB);
				else:
					part__ = [PB_];
				for PB__ in part__:
					part_.append(PB__);
			part = part_;
		return part;

	def randomselection(self, k=0, PB=None):
		if k == 0:
			return self.randomselection(1, self.box);
		if k > len(self.filter):
			return PB;

		PB_e = None;
		part = PB.mince(self.filter[k]);
		part = [PB_ for PB_ in part if PB_.colour == 0 and PB_.mass > 0];
		if len(part) == 0:
			return None;

		wt = [];
		for PB_ in part:
			wt.append(PB_.mass);
		wt_ = wt;
		len_ind = len(part);
		ind = range(0,len(part));
		while len_ind > 0:
			F = [0];
			m_tot = 0
			for i,m in enumerate(wt_):
				m_tot += m;
				F.append(m_tot);
			u = numpy.random.uniform(0, 1);
			m = m_tot*u;
			j = 0;
			for i in range(0, len(ind)):
				if m >= F[i+1]:
					continue;
				j = i;
				break;
			PB_ = part[ind[j]];
			PB_e = self.randomselection(k+1, PB_);
			if not PB_e is None:
				break;
			wt_ = [x for i,x in enumerate(wt) if not i == j];
			ind = [i for i in ind if not i == j];
			len_ind -= 1;

		return PB_e;

	def addrandomcell(self, bufferdim, n=1):
		if isinstance(bufferdim, (int, long, float)):
			bufferdim = [bufferdim for i in range(0, self.dim)];
		if not isinstance(bufferdim, list):
			bufferdim = [0 for i in range(0, self.dim)];

		self.buffer = bufferdim;

		if n == 0:
			return [];
		elif n > 1:
			cells = [];
			for i in range(n):
				cell = self.addrandomcell(bufferdim, 1);
				if not instanceof(cell, 'picassobox'):
					break;
				cells.append(cell);
			return cells;

		# Partition für zulässige Teile neu berechnen:
		self.filter = [];
		for PB in self.bounds + self.boxes:
			self.filter.append(PB.modify({'preshift': -bufferdim, 'colour': 1}));

		# Zufällige Selektion eines zulässigen Teils der Partition:
		PB_part = self.randomselection();
		if not isinstance(PB_part, 'picassobox'):
			return(None);

		# Zufällige Selektion eines Punkts in Box:
		x = PB_part.randomunif();
		sides = [[x[k], x[k]+h] for k, h in enumerate(bufferdim)];
		PB = picassobox(sides, 1);

		# Füge zum Netzwerk von Zellen:
		self.boxes.append(PB);

		return PB;

	def clonelist(self, x, deep=True):
		xx = x;
		if isinstance(x,list) or isinstance(x,tuple):
			xx = [];
			for y in x:
				yy = y;
				if deep:
					yy = self.clonelist(y, True);
				xx.append(yy);
			if isinstance(x,tuple):
				xx = tuple(xx);
		if isinstance(x,dict):
			xx = {};
			for key in x:
				y = x[key];
				yy = y;
				if deep:
					yy = self.clonelist(y, True);
				xx[key] = yy;
		return xx;

	pass;


class picassolines:
	def __init__(self, sides=None):
		if not sides is None:
			self.setframe(sides);
		self.edges = [];
		self.paths = [];
		self.points = [];
		self.radius = [];
		pass;

	def setframe(self, sides):
		sides = self.clonelist(sides);
		for i, x in enumerate(sides):
			if not isinstance(x, list):
				sides[i] = [0, x];
		self.dim = len(sides);
		self.box = picassobox(sides, 0);
		pass;

	def addpoints(self, x, r):
		self.points = x;
		if isinstance(r, list):
			r = [r for p in x];
		self.radius = r;
		pass;

	def addpointsfrompicassoboxes(self, PBs=[], r=1, relative=True):
		self.dim = PBs.dim;
		self.box = PBs.box;
		self.points = [PB.point(0.5) for PB in PBs.boxes];
		if relative:
			self.radius = [
				r*numpy.sqrt(
					sum([ ((x[1]-x[0])/2)**2 for x in PB.sides ])
				)
				for PB in PBs.boxes
			];
		else:
			self.radius = [r for PB in PBs.boxes];
		pass;
	
	def addpath(self, i, j, ptout=None, angleout=None, ptin=None, anglein=None, buffer=0, relative=True):
		self.edges.append([i, j]);
		radii = self.clonelist(self.radius);
		if relative:
			radii = [(1+buffer)*r for r in radii];
		else:
			radii = [r+buffer for r in radii];

		Pout = self.point[i];
		Rout = radii[i];
		Pin = self.point[j];
		Rin = radii[j];

		dP = [Pin[k]-Pout[k] for k in range(self.dim)];
		norm_dP = numpy.sqrt(sum([dx**2 for dx in dP]));
		u = [dx/norm_dP for dx in dP];
		sgn_dP = [numpy.sign(dx) for dx in dP];

		if ptout is None:
			ptout = [Rout*dx for dx in u];
		if ptin is None:
			ptin = [-Rin*dx for dx in u];
		if angleout is None:
			angleout = u;
		if anglein is None:
			anglein = [-dx for dx in u];

		path = [];

		## berechne Pfad...
		# unter Arbeit

		return(path);

	def clonelist(self, x, deep=True):
		xx = x;
		if isinstance(x,list) or isinstance(x,tuple):
			xx = [];
			for y in x:
				yy = y;
				if deep:
					yy = self.clonelist(y, True);
				xx.append(yy);
			if isinstance(x,tuple):
				xx = tuple(xx);
		if isinstance(x,dict):
			xx = {};
			for key in x:
				y = x[key];
				yy = y;
				if deep:
					yy = self.clonelist(y, True);
				xx[key] = yy;
		return xx;

	pass;
